File: leetcode-in-python/solution/162_find_peak_element.py
from typing import List
import unittest


class Solution:
    def findPeakElement(self, nums: List[int]) -> int:
        lo, hi = 0, len(nums)
        while 1 < hi - lo:  # 还剩一个元素
            mid = lo + (hi - lo) // 2
            if nums[mid - 1] > nums[mid]:
                hi = mid
            else:  # nums[mid] < nums[mid+1] or nums[mid] > nums[mid+1]
                lo = mid
        return lo


class TestSolution(unittest.TestCase):

    def setUp(self):
        self.test_case = [
            ([1, 2, 1, 3, 5, 6, 4], [1, 5]),
            ([1, 2, 3, 1], [2]),
        ]
        self.s = Solution()

    def test_solution(self):
        for nums, answer in self.test_case:
            ans = self.s.findPeakElement(nums)
            print(answer)
            self.assertIn(ans , answer)


if __name__ == '__main__':
    unittest.main()

File: leetcode-in-rust/src/solution/_162_find_peak_element.rs
pub struct Solution {}


impl Solution {
    pub fn find_peak_element(nums: Vec<i32>) -> i32 {
        let (mut lo, mut hi) = (0, nums.len());
        while 1 < hi - lo{
            let mid = lo + (hi - lo) / 2;
            if nums[mid-1] > nums[mid]{
                hi = mid
            } else {
                lo = mid
            }
        }
        lo as i32
    }
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_solution() {
        let test_case = vec![
            (vec![1, 2, 1, 3, 5, 6, 4], vec![1, 5]),
            (vec![1, 2, 3, 1], vec![2]),
            (vec![1, 2, 3], vec![2]),
            (vec![3, 2, 1], vec![0]),
            (vec![3], vec![0]),
            (vec![3, 2], vec![0]),
            (vec![2, 3], vec![1]),
        ];
        for (nums, answer) in test_case {
            let ans = Solution::find_peak_element(nums);
            assert_eq!(true, answer.contains(&ans));
        }
    }
}
